bench_test.go

```diff
@@ -0,0 +1,179 @@
+		}
+	})
+	b.StopTimer()
+}
+
+func benchmarkGet(b *testing.B, chunks, chunkSize, vkLen int) {
+	b.ResetTimer()
+
+	dir := b.TempDir()
+
+	store, metr, server := createServer(b, dir)
+	defer store.Close()
+	defer metr.Shutdown(context.Background())
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			mhs := randomMultihashes(b, chunks*chunkSize)
+			putMultihashes(b, mhs, chunks, vkLen, server, false)
+			rand.Shuffle(len(mhs), func(i, j int) {
+				mhs[i], mhs[j] = mhs[j], mhs[i]
+			})
+			getMultihashes(b, mhs, server, true)
+		}
+	})
+	b.StopTimer()
+}
+
+func putMultihashes(b *testing.B, mhs []multihash.Multihash, chunks, vkLen int, server *http.ServeMux, print bool) []multihash.Multihash {
+	chunkSize := len(mhs) / chunks
+
+	measurements := make([]time.Duration, 0, chunks)
+	for ichunk := 0; ichunk < chunks; ichunk++ {
+		start := time.Now()
+		istart := ichunk * chunkSize
+
+		body := dhstore.MergeIndexRequest{
+			Merges: make([]dhstore.Merge, 0, chunkSize),
+		}
+
+		vk := make([]byte, vkLen)
+		_, err := crand.Read(vk)
+		require.NoError(b, err)
+
+		for imh := istart; imh < istart+chunkSize; imh++ {
+			body.Merges = append(body.Merges, dhstore.Merge{
+				Key:   mhs[imh],
+				Value: vk,
+			})
+		}
+
+		reqBody, err := json.Marshal(body)
+		require.NoError(b, err)
+		resp := httptest.NewRecorder()
+		req := httptest.NewRequest(http.MethodPut, "/multihash", bytes.NewBuffer(reqBody))
+		server.ServeHTTP(resp, req)
+
+		respBody, err := io.ReadAll(resp.Body)
+		require.NoError(b, err)
+		require.Equal(b, http.StatusAccepted, resp.Result().StatusCode, string(respBody))
+		measurements = append(measurements, time.Since(start))
+	}
+
+	if print {
+		printBenchmark("PUT benchmark", measurements)
+	}
+	return mhs
+}
+
+func getMultihashes(b *testing.B, mhs []multihash.Multihash, server *http.ServeMux, print bool) {
+	measurements := make([]time.Duration, 0, len(mhs))
+	for _, mh := range mhs {
+		start := time.Now()
+		resp := httptest.NewRecorder()
+		req := httptest.NewRequest(http.MethodGet, "/multihash/"+mh.B58String(), nil)
+		server.ServeHTTP(resp, req)
+
+		respBody, err := io.ReadAll(resp.Body)
+		require.NoError(b, err)
+		require.Equal(b, http.StatusOK, resp.Result().StatusCode, string(respBody))
+		measurements = append(measurements, time.Since(start))
+	}
+	if print {
+		printBenchmark("GET benchmark", measurements)
+	}
+}
+
+func printBenchmark(name string, m []time.Duration) {
+	var total int64
+	var max int64
+
+	for _, d := range m {
+		total += int64(d)
+		if int64(d) > max {
+			max = int64(d)
+		}
+	}
+
+	mean := int64(total) / int64(len(m))
+	var deviation int64
+
+	for _, d := range m {
+		deviation += int64(math.Pow(float64(int64(d)-mean), 2))
+	}
+	deviation = int64(math.Sqrt(float64(deviation / int64(len(m)))))
+
+	log.Print(name)
+	log.Printf("\tmean: %v", time.Duration(mean))
+	log.Printf("\tstdev: %v", time.Duration(deviation))
+	log.Printf("\tmax: %v", time.Duration(max))
```

Comment:
* benchmark default printing already gives us measurements for these. do we really need this?
* use `b.Log` for printing stuff.

Reply:
Changed to b.Log. Benchmark gives us stats per *run* while we are interested in stats per query.

Reply:
🤔 per query?

If there is something you want to report that's not already covered consider using `b.ReportMetric` instead.

Reply:
good point, changed to `ReportMetric`